### ISSUE TITLE
ParamSpec/Type in Type[Callable[P,T]] binds to params of __init__, not __call__ for callable classes.

### ISSUE THREAD

**Issue by gridzero** [microsoft/pyright#5546]
**Describe the bug**

I'm writing a higher-level generic function which takes a parameter which is the _type_ of some callable, and returns a result based on the signature of that callable; i.e.:

```python
def buggy_fn(fn_t: type[Callable[P,T]]) -> list[Callable[P,T]]:
    ...
```

So, fn_t is a _type_, and _instances_ of that type are callable with signature `some_instance(*args: P.args, **kwargs: P.kwargs) -> T`

This works with the expected bindings of P and T when passing arguments of `Callable[[CorrectArgs],ResType]` and `type(actual_fn)`, but when passed a class object for some class implementing `__call__`, the paramspec and return type don't bind to those of the `__call__` method as expected,  but to those of the `__init__` method (even when inheriting the default `__init__` from object).

The bindings that are seen would be correct if the parameter annotation was just `Callable[P,T]` (since the class object _is_ callable with the effective signature of `__init__`), but not for `type[Callable[P,T]]` (as _instances_ are not callable with `__init__`'s  signature.)

**To Reproduce**

* Define class implementing `__call__` (and optionally, `__init__`)
* Pass that class object to some function expecting `type[Callable[P,T]]` (or `typing.Type[Callable[P,T]]`)

**Expected behavior**
P and T are bound to the parameters and return type of the class's `__call__` method (minus the leading self)

**Observed behaviour**
P is bound to the parameters of the class's `__init__` method (minus the leading self), and T to the argument itself, mirroring the effective signature of the class constructor.

**Code or Screenshots**

```python
from __future__ import annotations

from collections.abc import Callable
from typing import ParamSpec, TypeVar

T = TypeVar("T")
P = ParamSpec("P")

class CorrectArg: pass
class IncorrectArg: pass
class ResultType: pass

class ProblemClass:
    def __init__(self, x:IncorrectArg)->None: pass
    def __call__(self, x:CorrectArg)->ResultType: return ResultType()

def buggy_fn(fn_t: type[Callable[P,T]]) -> list[Callable[P,T]]:
    return []

# Used as testdata for illustrating correct behaviour
def example_actual_fn(x:CorrectArg)->ResultType: return ResultType()

r1 = buggy_fn(type(example_actual_fn))
reveal_type(r1)  
# Pylance Correct: Type of "r1" is "list[(x: CorrectArg) -> ResultType]"
# MyPy Correct: list[def (x: CorrectArg) -> ResultType]

r2 = buggy_fn(Callable[[CorrectArg], ResultType])
reveal_type(r2)
# Pylance Correct: Type of "r2" is "list[(CorrectArg) -> ResultType]"
# MyPy incorrect: list[def (*Any, **Any) -> Any]
# MyPy issue: https://github.com/python/mypy/issues/15593

r3 = buggy_fn(ProblemClass)
reveal_type(r3)
# Pylance INCORRECT: Type of "r3" is "list[(x: IncorrectArg) -> ProblemClass]"
# MyPy correct: list[def (x: CorrectArg) -> ResultType]

```

**VS Code extension or command-line**
VSCode Extension v2023.12.0

**Additional context**
Mypy gets this case correct, but does have it's own bug noted above when directly passing Callable[[T1,T2],R] as an argument.

**Comment by erictraut** [comment]
You're using an invalid type annotation here. The `type` annotation does not support `Callable`. it supports only types that can be instantiated. A `Callable` (in particular, a function) cannot be instantiated — at least not in a manner that preserves its signature.

```python
def func(): return []

x = type(func)
print(x)

y = x()  # Runtime error
```

Pyright (and mypy) should probably emit an error for `type[Callable[...]]`. This is the first time I've ever seen someone attempt to use that, so it's likely that it's simply never come up before.

**Comment by erictraut** [comment]
I've updated pyright to disallow the use of `type[Callable]` (and I've recommended that mypy do the same). This will be included in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.319, which I just published. It will also be included in a future release of pylance.

**Comment by cainmagi** [comment]
Thank you! Once I update `Pylance`, I will try to check whether this issue still exists or not.
